File: java/src/main/java/com/github/algo/graph/MaxAreaOfIsland.java
package com.github.algo.graph;

public class MaxAreaOfIsland {

    int[][] movements = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    private int dfs(int i, int j, int[][] grid) {
        grid[i][j] = -1;
        int count = 1;
        for (int[] move : movements) {
            int di = i + move[0], dj = j + move[1];
            if (di < grid.length && di >= 0 && dj < grid[di].length && dj >= 0) {
                if (grid[di][dj] == 1) {
                    count += dfs(di, dj, grid);
                }
            }
        }
        return count;
    }

    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 1) {
                    int area = dfs(i, j, grid);
                    maxArea = Math.max(area, maxArea);
                }
            }
        }
        return maxArea;
    }

}

File: python/main/graphs/MaxAreaOfIsland.py
from typing import List

movement = [(-1, 0), (1, 0), (0, 1), (0, -1)]


class Solution:

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        maxSize = 0

        def dfs(grid, i, j):
            count = 1
            grid[i][j] = -1
            for m in movement:
                xi, xj = i + m[0], j + m[1]
                if 0 <= xi < len(grid) and 0 <= xj < len(grid[i]):
                    if grid[xi][xj] == 1:
                        count += dfs(grid, xi, xj)
            return count

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 1:
                    size = dfs(grid, i, j)
                    maxSize = max(maxSize, size)
        return maxSize


if __name__ == "__main__":
    solution = Solution()
    output = solution.maxAreaOfIsland([[0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
                                       [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0],
                                       [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
                                       [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
                                       [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0]])
    print(output)
